packages/@aws-cdk/aws-s3/test/test.auto-delete-objects.ts

```diff
@@ -0,0 +1,218 @@
+      const s3Mock = newS3ClientMock();
+
+      const event: Partial<AWSLambda.CloudFormationCustomResourceCreateEvent> = {
+        RequestType: 'Create',
+        ResourceProperties: {
+          ServiceToken: 'Foo',
+          BucketName: 'MyBucket',
+        },
+      };
+      await invokeHandler(event, s3Mock);
+
+      sinon.assert.notCalled(s3Mock.listObjectVersions);
+      sinon.assert.notCalled(s3Mock.deleteObjects);
+
+      test.done();
+    },
+
+    async 'does nothing on update event'(test: Test) {
+      sinon.reset();
+
+      const s3Mock = newS3ClientMock();
+
+      const event: Partial<AWSLambda.CloudFormationCustomResourceUpdateEvent> = {
+        RequestType: 'Update',
+        ResourceProperties: {
+          ServiceToken: 'Foo',
+          BucketName: 'MyBucket',
+        },
+      };
+      await invokeHandler(event, s3Mock);
+
+      sinon.assert.notCalled(s3Mock.listObjectVersions);
+      sinon.assert.notCalled(s3Mock.deleteObjects);
+
+      test.done();
+    },
+
+    async 'deletes no objects on delete event when bucket has no objects'(test: Test) {
+      sinon.reset();
+
+      const listObjectVersionsMock = sinon.fake.returns({ Versions: [] });
+      const s3Mock = newS3ClientMock(listObjectVersionsMock);
+
+      const event: Partial<AWSLambda.CloudFormationCustomResourceDeleteEvent> = {
+        RequestType: 'Delete',
+        ResourceProperties: {
+          ServiceToken: 'Foo',
+          BucketName: 'MyBucket',
+        },
+      };
+      await invokeHandler(event, s3Mock);
+
+      sinon.assert.calledOnce(s3Mock.listObjectVersions);
+      sinon.assert.notCalled(s3Mock.deleteObjects);
+
+      test.done();
+    },
+
+    async 'deletes all objects on delete event'(test: Test) {
+      sinon.reset();
+
+      const listObjectVersionsMock = sinon.fake.returns({
+        Versions: [
+          { Key: 'Key1', VersionId: 'VersionId1' },
+          { Key: 'Key2', VersionId: 'VersionId2' },
+        ],
+      });
+      const s3Mock = newS3ClientMock(listObjectVersionsMock);
+
+      const event: Partial<AWSLambda.CloudFormationCustomResourceDeleteEvent> = {
+        RequestType: 'Delete',
+        ResourceProperties: {
+          ServiceToken: 'Foo',
+          BucketName: 'MyBucket',
+        },
+      };
+      await invokeHandler(event, s3Mock);
+
+      sinon.assert.calledOnce(s3Mock.listObjectVersions);
+      sinon.assert.calledOnce(s3Mock.deleteObjects);
+
+      test.done();
+    },
+  },
+};
+
+async function invokeHandler(event: Partial<AWSLambda.CloudFormationCustomResourceEvent>, s3client: any) {
+  return handler(event as any, s3client);
+}
+
+function newS3ClientMock(listObjectVersionsMock?: any, deleteObjectsMock?: any) {
+  return {
+    listObjectVersions: sinon.stub().returns({
+      promise: listObjectVersionsMock ?? sinon.fake.returns({}),
+    }),
+    deleteObjects: sinon.stub().returns({
+      promise: deleteObjectsMock ?? sinon.fake.returns({}),
+    }),
+  };
+}
```

Comment:
Could we see if we can avoid passing the `s3Client` to handler (in case we are doing it just for testing) and rely on mocking the request calls?